File: java/memento/src/main/Book.java
package main;

import java.time.LocalDateTime;

public class Book {
    private String title;
    private String author;
    private String isbn;
    private LocalDateTime lastEdited;

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
        setLastEdited();
    }

    public String getAuthor() {
        return author;
    }

    public void setAuthor(String author) {
        this.author = author;
        setLastEdited();
    }

    public String getIsbn() {
        return isbn;
    }

    public void setIsbn(String isbn) {
        this.isbn = isbn;
        setLastEdited();
    }

    public LocalDateTime getLastEdited() {
        return lastEdited;
    }

    public void setLastEdited() {
        this.lastEdited = LocalDateTime.now();
    }

    public Memento createUndo()
    {
        return new Memento(title, author, isbn, lastEdited);
    }

    public void restoreFromUndo(Memento memento)
    {
        title = memento.getTitle();
        author = memento.getAuthor();
        isbn = memento.getIsbn();
        lastEdited = memento.getLastEdited();
    }

    @Override
    public String toString() {
        return this.isbn + " " + this.author + " " + this.title + " edited: " + this.lastEdited;
    }
}

File: csharp/Memento/Book.cs
namespace Memento;

public class Book
{
    private string _title;
    private string _author;
    private string _isbn;
    private DateTime _lastEdited;

    public string Title
    {
        get { return _title; }
        set
        {
            _title = value;
            SetLastEdited();
        }
    }
    public string Author
    {
        get { return _author; }
        set
        {
            _author = value;
            SetLastEdited();
        }
    }
    public string Isbn
    {
        get { return _isbn; }
        set
        {
            _isbn = value;
            SetLastEdited();
        }
    }

    private void SetLastEdited()
    {
        _lastEdited = DateTime.UtcNow;
    }

    public Memento CreateUndo()
    {
        return new Memento(_isbn, _title, _author, _lastEdited);
    }

    public void RestoreFromUndo(Memento memento)
    {
        _title = memento.Title;
        _author = memento.Author;
        _lastEdited = memento.LastEdited;
        _isbn = memento.Isbn;
    }

    public void ShowBook()
    {
        Console.WriteLine("{0}, {1}, {2},  edited: {3}", _isbn, _author, _title, _lastEdited);
    }
}